workflow/snakemake_rules/remote_files.smk

```diff
@@ -0,0 +1,109 @@
+"""
+Shared functions to transparently handle local and remote inputs/outputs.
+
+Snakemake itself tries to offer a similar sort of scheme-level transparency for
+remote files through its ``AUTO`` provider, but it's broken.
+
+The ``AUTO`` provider enumerates and loads all the various
+``snakemake.remote.*`` modules on each call to ``AUTO.remote()``.  This slows
+things down unnecessarily, though isn't a deal breaker.  However, during
+enumeration, it protects against unsupported modules due to missing deps, but
+*not* supported modules which fail to initialize because of missing required
+credentials/config from the environment or missing required parameters to the
+remote provider constructor or ``AUTO.remote()``.  This alone makes it
+completely unusable.  Even with a usable ``AUTO``, we'd still need some
+additional handling for local paths.
+
+We could make a better functioning ``AUTO`` equivalent ourselves by similarly
+building the scheme → remote function mapping dynamically, but it didn't seem
+worth it to support use cases that aren't concrete right now (http(s), s3, gs
+are all known use cases for us).  The implementation here could certainly be
+further developed in that direction in the future, however.
+"""
+from functools import partial
+from importlib import import_module
+from urllib.parse import urlsplit, urljoin
+
+
+def _remote_producer(init):
+    """
+    Produce a remote input function by loading a :class:`snakemake.remote`
+    module and passing the module to an *init* function.
+
+    The module is named by the first argument of the *init* function.  *init*
+    will be called with the loaded module and is expected to return a remote
+    input function suitable for use inside :func:`path_or_url`.
+
+    Returns a "producer" function which when called with no arguments returns
+    the remote input function.
+
+    The call to *init* (that is, the production of the remote input function)
+    is deferred until the first time the producer is called.
+
+    If the remote module is not available because required dependencies aren't
+    installed, then a :exc:`WorkflowError` exception object is returned on the
+    first call of the producer.
+    """
+    # Name of first argument to *init*.
+    remote_name = init.__code__.co_varnames[0]
```

Comment:
What if we defined `remote_name` as an argument to `_remote_producer` instead of using code introspection like this? I mostly was surprised to see a variable name used as a variable itself and wonder if Future Me (or others) will be similarly confused later.

Reply:
I considered that exact `remote_name` argument, but to me, small, contained uses of introspection like this to make tiny lambdas more scannable seems like a perfect use case. I realize it's not a widely-used technique in most Python in the wild, though. I hoped the documentation and comment would abate any future confusion.

Here's what it'd look like to add a `remote_name` argument. It seems to me that it takes the same quality and quantity of effort to understand what's happening as with the introspection version, but I'm maybe bias here.

```diff
diff --git a/workflow/snakemake_rules/remote_files.smk b/workflow/snakemake_rules/remote_files.smk
index 2e143d9a..a8fb49b0 100644
--- a/workflow/snakemake_rules/remote_files.smk
+++ b/workflow/snakemake_rules/remote_files.smk
@@ -25,14 +25,13 @@ from importlib import import_module
 from urllib.parse import urlsplit, urljoin
 
 
-def _remote_producer(init):
+def _remote_producer(remote_name, init):
     """
-    Produce a remote input function by loading a :class:`snakemake.remote`
-    module and passing the module to an *init* function.
+    Produce a remote input function by loading the :class:`snakemake.remote`
+    module named by *remote_name* and passing the module to an *init* function.
 
-    The module is named by the first argument of the *init* function.  *init*
-    will be called with the loaded module and is expected to return a remote
-    input function suitable for use inside :func:`path_or_url`.
+    *init* will be called with the loaded module and is expected to return a
+    remote input function suitable for use inside :func:`path_or_url`.
 
     Returns a "producer" function which when called with no arguments returns
     the remote input function.
@@ -45,7 +44,6 @@ def _remote_producer(init):
     first call of the producer.
     """
     # Name of first argument to *init*.
-    remote_name = init.__code__.co_varnames[0]
     remote = None
 
     def producer():
@@ -58,10 +56,10 @@ def _remote_producer(init):
 
 
 SCHEME_REMOTES = {
-    "https": _remote_producer(lambda HTTP: HTTP.RemoteProvider().remote),
-    "http": _remote_producer(lambda HTTP: partial(HTTP.RemoteProvider().remote, insecure = True)),
-    "s3": _remote_producer(lambda S3: S3.RemoteProvider().remote),
-    "gs": _remote_producer(lambda GS: GS.RemoteProvider().remote),
+    "https": _remote_producer("HTTP", lambda HTTP: HTTP.RemoteProvider().remote),
+    "http": _remote_producer("HTTP", lambda HTTP: partial(HTTP.RemoteProvider().remote, insecure = True)),
+    "s3": _remote_producer("S3", lambda S3: S3.RemoteProvider().remote),
+    "gs": _remote_producer("GS", lambda GS: GS.RemoteProvider().remote),
     "": lambda: lambda local_path, **kwargs: local_path,
 }
```

Reply:
In any case, happy to commit the `remote_name` version above if you think it's better.

Reply:
Thank you for placing these side by side! From this view, the explicit remote argument is slightly easier to interpret, because it is clearer to me how that value is being used as a variable.

Reply:
Got it. Repushed to make this change.